Improve asset ingestion

bigkahuna443 opened the pull request:
1. Add `AssetTypeNonConflict` for mods to subclass from. ModAssets with this base type won't log conflicts.
2. Extend blacklist to avoid ingesting Ahorn and Loenn files as well as common code/documentation files.
3. Refactor `GuessType` to allow custom mod types to take precedence over generic types.

bigkahuna443 committed:
Add Loenn files and Decal Registry XMLs to the non-conflict list

We don't care about files in Loenn or DecalRegistry.xml files having the same name, so add both to the non-conflict list

bigkahuna443 committed:
Whitespace

microlith57 commented on Celeste.Mod.mm/Mod/Everest/Everest.Content.cs:
```diff
@@ -798,9 +801,12 @@ public static void Add(string path, ModAsset metadata)
 
                 } else if (file.StartsWith("Ahorn/")) {
-                    // Special case: Fallback type for anything inside of the Ahorn folder.
+                    // Fallback types for anything inside of the Ahorn or Loenn folders.
                     // Will be ignored during collision checks.
```
nitpick: this comment applies to both `else if` branches but is inside of one of them

bigkahuna443 replied:
Yeah auto-formatter wasn't liking the comment position but I do like it better outside lol, fixed

bigkahuna443 committed:
Swap comment location

coloursofnoise commented:
Another one I see regularly is the backup `everest.yaml` created by the Ahorn Dependency Generator.

It may also be beneficial to expose an endpoint for mods to register their own no-conflict paths, such as the CollabUtils or SkinModHelper config files.

bigkahuna443 commented:
> It may also be beneficial to expose an endpoint for mods to register their own no-conflict paths, such as the CollabUtils or SkinModHelper config files.

Yeah that's a good point, I noticed that you could register an OnGuessType delegate but not label it as "non-conflict". And the delegate has lower priority than all the other types, so e.g. a custom YAML asset type wouldn't even get registered. I have an idea on how to improve both of those I think

Kalobi commented:
I looked into this a while ago, but got distracted by school. My idea was a .everestignore file that works like a .gitignore and lets people exclude files from being scanned by Everest, probably using [DotIgnore](https://github.com/markashleybell/MAB.DotIgnore). I guess that would only work for things like documentation though, not for things that you want to be scanned, but prevent from conflicting.

bigkahuna443 commented:
Updated:

This change makes assets *not* log conflicts by default, which seems more reasonable than filtering everything that gets ingested.

Added `AssetTypeConflict` for existing conflict types + specifying user-specified types they want to log conflicts for.

Removed `AssetTypeAhorn` and `NonConflictTypes` as they aren't relevant with the new change. Couldn't find either in use anywhere.

Moved custom asset type resolution before generic asset types. Improved formatting in the section.

bigkahuna443 committed:
Re-organize asset types

This changes makes assets *not* log conflicts by default, which seems more reasonable than going through weeding through everything that gets ingested.

Added `AssetTypeConflict` for existing conflict types + user-specified types they want to log conflicts for.

Removed `AssetTypeAhorn` and `NonConflictTypes` as they aren't relevant with the new change. Couldn't find it being used elsewhere.

Moved custom asset type resolution before generic asset types. Improved formatting in the section.

bigkahuna443 committed:
Make lua/text non-conflict

Pulled out specific unique files as their own types so that we can make the generic types non-conflict

XMLs need it for tilesets/etc, which can be named anything, and it's probably not safe to do for YAML files :(

bigkahuna443 committed:
Revert + improve blacklist

Went too aggressive on the last commit, defaulting to no-log conflicts may not be safe without a content rework.

bigkahuna443 committed:
Move filename check to root folder

These are limited to documentation so makes sense to move here (plus most likely place for conflicts)

Brokemia commented:
I noticed you're removing the `.git` blacklist. Is there a reason behind that?

Brokemia commented:
Ignore me, it's because you're blacklisting everything starting with `.` anyway
